storage/innobase/os/os0file.cc

```diff
@@ -3084,168 +3080,53 @@ static void write_io_callback(void *c)
-				" native AIO support.";
-
-			int ret = io_destroy(io_ctx);
-			ut_a(ret != -EINVAL);
-			ut_ad(ret != -EFAULT);
-
-			return(false);
-		}
-	}
-	else {
-		fd = my_open(log_file_path.c_str(), O_RDONLY | O_CLOEXEC,
-			     MYF(0));
-
-		if (fd == -1) {
-
-			ib::warn() << "Unable to open \"" << log_file_path
-				   << "\" to check native"
-				   << " AIO read support.";
-
-			int ret = io_destroy(io_ctx);
-			ut_a(ret != EINVAL);
-			ut_ad(ret != EFAULT);
-
-			return(false);
-		}
-	}
-
-	struct io_event	io_event;
-
-	memset(&io_event, 0x0, sizeof(io_event));
-
-	byte* ptr = static_cast<byte*>(aligned_malloc(srv_page_size,
-						      srv_page_size));
-
-	struct iocb	iocb;
-
-	/* Suppress valgrind warning. */
-	memset(ptr, 0, srv_page_size);
-	memset(&iocb, 0x0, sizeof(iocb));
-
-	struct iocb* p_iocb = &iocb;
-
-	if (!srv_read_only_mode) {
-
-		io_prep_pwrite(p_iocb, fd, ptr, srv_page_size, 0);
-
-	}
-	else {
-		ut_a(srv_page_size >= 512);
-		io_prep_pread(p_iocb, fd, ptr, 512, 0);
-	}
-
-	int	err = io_submit(io_ctx, 1, &p_iocb);
-
-	if (err >= 1) {
-		/* Now collect the submitted IO request. */
-		err = io_getevents(io_ctx, 1, 1, &io_event, NULL);
-	}
-
-	aligned_free(ptr);
-	my_close(fd, MYF(MY_WME));
-
-	switch (err) {
-	case 1:
-		{
-			int ret = io_destroy(io_ctx);
-			ut_a(ret != -EINVAL);
-			ut_ad(ret != -EFAULT);
-
-			return(true);
-		}
-
-	case -EINVAL:
-	case -ENOSYS:
-		ib::warn()
-			<< "Linux Native AIO not supported. You can either"
-			" move "
-			<< (srv_read_only_mode ? log_file_path : "tmpdir")
-			<< " to a file system that supports native"
-			" AIO or you can set innodb_use_native_aio to"
-			" FALSE to avoid this message.";
-
-		/* fall through. */
-	default:
-		ib::warn()
-			<< "Linux Native AIO check on "
-			<< (srv_read_only_mode ? log_file_path : "tmpdir")
-			<< "returned error[" << -err << "]";
-	}
-
-	int ret = io_destroy(io_ctx);
-	ut_a(ret != -EINVAL);
-	ut_ad(ret != -EFAULT);
-
-	return(false);
-}
-#endif
-
 int os_aio_init() noexcept
 {
   int max_write_events= int(srv_n_write_io_threads *
                             OS_AIO_N_PENDING_IOS_PER_THREAD);
   int max_read_events= int(srv_n_read_io_threads *
                            OS_AIO_N_PENDING_IOS_PER_THREAD);
   int max_events= max_read_events + max_write_events;
-  int ret;
-#if LINUX_NATIVE_AIO
-  if (srv_use_native_aio && !is_linux_native_aio_supported())
-    goto disable;
-#endif
-
-  ret= srv_thread_pool->configure_aio(srv_use_native_aio, max_events);
+  int ret= 1;
 
-#ifdef LINUX_NATIVE_AIO
-  if (ret)
+#if defined __linux__ && (defined HAVE_URING || defined LINUX_NATIVE_AIO)
+  if (srv_use_native_aio)
   {
-    ut_ad(srv_use_native_aio);
-disable:
-    ib::warn() << "Linux Native AIO disabled.";
-    srv_use_native_aio= false;
-    ret= srv_thread_pool->configure_aio(false, max_events);
-  }
+    switch (srv_linux_aio_method) {
+    case SRV_LINUX_AIO_AUTO:
+    case SRV_LINUX_AIO_IO_URING:
+#ifdef HAVE_URING
```

Comment:
I'd prefer HAVE_URING and  HAVE_LINUX_NATIVE_AIO to be contained inside tpool and not used outside of it.
my_printf_error  with ME_WARNING inside tpool is used by uring  already, though I think fprintf is less dependencies. if you need some reporting

Reply:
Only on Linux, innodb_use_native_aio=ON *can* fail in the first place. You do not need to apply logic for other OSes at all, even if introduce it, it won't change anything .